src/ompl/geometric/planners/rrt/src/RRTstarConnect.cpp

```diff
@@ -0,0 +1,1260 @@
+
+	tree->add(motion);
+	tgi.xmotion = motion;
+	motion->parent->children.push_back(motion);
+
+	// update existing motions
+	for (std::size_t i = 0; i < nbh.size(); ++i)
+	{
+		if (nbh[i] != motion->parent)
+		{
+			base::Cost nbhIncCost;
+			if (opt_->isSymmetric())
+				nbhIncCost = incCosts[i];
+			else
+				nbhIncCost = opt_->motionCost(motion->state, nbh[i]->state);
+			base::Cost nbhNewCost = opt_->combineCosts(motion->cost, nbhIncCost);
+			if (opt_->isCostBetterThan(nbhNewCost, nbh[i]->cost))
+			{
+				if (valid[i] == 0)
+				{
+					validMotion = (!useKNearest_ || si_->distance(nbh[i]->state, motion->state) < maxDistance_) &&
+							(tgi.start ? si_->checkMotion(motion->state, nbh[i]->state) :
+									si_->isValid(motion->state) && si_->checkMotion(nbh[i]->state, motion->state));
+
+				}
+				else
+				{
+					validMotion = (valid[i] == 1);
+				}
+
+				if (validMotion)
+				{
+					// Remove this node from its parent list
+					removeFromParent(nbh[i]);
+
+					// Add this node to the new parent
+					nbh[i]->parent = motion;
+					nbh[i]->incCost = nbhIncCost;
+					nbh[i]->cost = nbhNewCost;
+					nbh[i]->parent->children.push_back(nbh[i]);
+
+					// Update the costs of the node's children
+					updateChildCosts(nbh[i]);
+
+					checkForSolution = true;
+				}
+			}
+		}
+	}
+
+
+    return reach ? REACHED : ADVANCED;
+}
+
+ompl::base::PlannerStatus ompl::geometric::RRTstarConnect::solve(const base::PlannerTerminationCondition &ptc)
+{
+    checkValidity();
+    auto *goal = dynamic_cast<base::GoalSampleableRegion *>(pdef_->getGoal().get());
+
+    if (goal == nullptr)
+    {
+        OMPL_ERROR("%s: Unknown type of goal", getName().c_str());
+        return base::PlannerStatus::UNRECOGNIZED_GOAL_TYPE;
+    }
+
+    while (const base::State *st = pis_.nextStart())
+    {
+        auto *motion = new Motion(si_);
+        si_->copyState(motion->state, st);
+        motion->root = motion->state;
+        tStart_->add(motion);
+        startMotions_.push_back(motion);
+    }
+
+    // And assure that, if we're using an informed sampler, it's reset
+    infSampler_.reset();
+
+    if (tStart_->size() == 0)
+    {
+        OMPL_ERROR("%s: Motion planning start tree could not be initialized!", getName().c_str());
+        return base::PlannerStatus::INVALID_START;
+    }
+
+    if (!goal->couldSample())
+    {
+        OMPL_ERROR("%s: Insufficient states in sampleable goal region", getName().c_str());
+        return base::PlannerStatus::INVALID_GOAL;
+    }
+
+    // Allocate a sampler if necessary
+    if (!sampler_ && !infSampler_)
+    {
+        allocSampler();
+    }
+
+    OMPL_INFORM("%s: Starting planning with %d states already in datastructure", getName().c_str(),
+                (int)(tStart_->size() + tGoal_->size()));
+
+    TreeGrowingInfo tgi;
+    tgi.xstate = si_->allocState();
+
+    Motion *approxsol = nullptr;
+    double approxdif = std::numeric_limits<double>::infinity();
+    auto *rmotion = new Motion(si_);
```

Comment:
Hi

Thank you very much for the implementation! It works great.

Just one small note that this line here causes small memory leaks that I discovered using Valgring Memcheck (and CLion).

To avoid that, you simply have to free the state right before the `delete(rmotion)` line further down this file.

Reply:
Hi Ignat
I appreciate your review. 
I have already implemented the fix for the memory leak in the latest commit.